arviz/kde_utils.py

```diff
@@ -0,0 +1,812 @@
+
+    return output
+
+
+def _fixed_point(t, N, k_sq, a_sq):
+    """
+    Implementation of the function t-zeta*gamma^[l](t) derived
+    from equation (30) in [1]
+
+    References
+    ----------
+    .. [1] Kernel density estimation via diffusion.
+       Z. I. Botev, J. F. Grotowski, and D. P. Kroese.
+       Ann. Statist. 38 (2010), no. 5, 2916--2957.
+    """
+
+    k_sq = np.asfarray(k_sq, dtype=np.float64)
+    a_sq = np.asfarray(a_sq, dtype=np.float64)
+
+    l = 7
+    f = np.sum(np.power(k_sq, l) * a_sq * np.exp(-k_sq * np.pi ** 2 * t))
+    f *= 0.5 * np.pi ** (2.0 * l)
+
+    for j in reversed(range(2, l)):
+        c1 = (1 + 0.5 ** (j + 0.5)) / 3
+        c2 = np.product(np.arange(1.0, 2 * j + 1, 2, dtype=np.float64))
+        c2 /= (np.pi / 2) ** 0.5
+        t_j = np.power((c1 * (c2 / (N * f))), (2.0 / (3.0 + 2.0 * j)))
+        f = np.sum(k_sq ** j * a_sq * np.exp(-k_sq * np.pi ** 2.0 * t_j))
+        f *= 0.5 * np.pi ** (2 * j)
+
+    out = t - (2 * N * np.pi ** 0.5 * f) ** (-0.4)
+    return out
+
+
+def _root(function, N, args, x):
+    # The idea here was borrowed from KDEpy implementation.
+    # The right bound is at most 0.01
+    found = 0
+    N = max(min(1050, N), 50)
+    tol = 10e-12 + 0.01 * (N - 50) / 1000
+
+    while found == 0:
+        try:
+            bw, res = brentq(function, 0, 0.01, args=args, full_output=True,
+                             disp=False)
+            found = 1 if res.converged else 0
+        except ValueError:
+            bw = 0
+            tol *= 2.0
+            found = 0
+        if bw <= 0:
+            warnings.warn(
+                "Improved Sheather-Jones did not converge to a positive value. "
+                "Using Silverman's rule instead.",
+                Warning
+            )
+            bw = (bw_silverman(x) / np.ptp(x)) ** 2
+            return bw
+        if tol >= 1:
+            warnings.warn(
+                "Improved Sheather-Jones did not converge. "
+                "Using Silverman's rule instead.",
+                Warning
+            )
+            bw = (bw_silverman(x) / np.ptp(x)) ** 2
+    return bw
+
+# KDE Utilities ---------------------------------------------------------------------
+
+def check_type(x):
+    """
+    Checks the input is of the correct type.
+    It only accepts numeric lists/numpy arrays of 1 dimension.
+    If input is not of the correct type, an informative message is thrown.
+
+    Parameters
+    ----------
+    x : Object whose type is checked before computing the KDE.
+
+    Returns
+    -------
+    x : 1-D numpy array
+        If no error is thrown, a 1 dimensional array of
+        sample data from the variable for which a density estimate is desired.
+
+    """
+    if not isinstance(x, (list, np.ndarray)):
+        raise ValueError((
+            f"`x` is of the wrong type.\n"
+            f"Can't produce a density estimator for {type(x)}.\n"
+            f"Please input a numeric list or numpy array."
```

Comment:
what about a list with numbers?

Reply:
I think I'm missing something. The current implementation does accept a numeric list (which is a list with numbers, isn't it?). Do you mean something different?

Reply:
Sorry for being unclear. I suggest to write f"Please input a list with numbers or numpy array."